src/mesh/interpolation/hermite_spline_z.cxx

```diff
@@ -125,32 +142,37 @@ Field3D ZHermiteSpline::interpolate(const Field3D& f, const std::string& region)
                                     ? y_offset == 0 ? "RGN_NOBNDRY" : "RGN_NOX"
                                     : "RGN_ALL";
   Field3D fz = bout::derivatives::index::DDZ(f, CELL_DEFAULT, "DEFAULT", fz_region);
-  localmesh->communicateXZ(fz);
-  // communicate in y, but do not calculate parallel slices
-  {
+  if (region != "RGN_NOBNDRY" and region != "RGN_NOX") {
+    localmesh->communicateXZ(fz);
+  }
+  if (y_offset != 0 or (region != "RGN_NOBNDRY" and region != "RGN_NOY")) {
+    // communicate in y, but do not calculate parallel slices
     auto h = localmesh->sendY(fz);
     localmesh->wait(h);
   }
 
+  const int ncz = localmesh->LocalNz;
+
   BOUT_FOR(i, f.getRegion(region)) {
     const int x = i.x();
     const int y = i.y();
     const int z = i.z();
 
-    if (skip_mask(x, y, z))
-      continue;
+    if (with_mask) {
+      if (skip_mask(x, y, z))
+        continue;
+    }
 
     // Due to lack of guard cells in z-direction, we need to ensure z-index
     // wraps around
-    const int ncz = localmesh->LocalNz;
-    const int z_mod = ((k_corner(x, y, z) % ncz) + ncz) % ncz;
-    const int z_mod_p1 = (z_mod + 1) % ncz;
+    const int z_mod = k_corner[i.ind];
+    const int z_mod_p1 = (z_mod < ncz - 1) ? (z_mod + 1) : 0;
```

Comment:
What if made `k_corner` an `Array<Ind3D>`?

```suggestion
    const auto z_mod = k_corner[i.ind];
    const auto z_mod_p1 = z_mod.zp();
```

Similarly for `y_next`. Then `f_interp` calculation could be done entirely with single-indices

I think this moves some complexity into the construction of `k_corner`, but that's set-up cost so one-off.